internal/util/resource/resource.go

```diff
@@ -76,28 +80,30 @@ func Track(obj any, token *Token) {
 	// because otherwise profile will hold a reference to obj and finalizer will never run
 	p.Add(token, 1)
 
-	var stack string
+	msg := fmt.Sprintf("%T has not been finalized", obj)
 	if devbuild.Enabled {
-		stack = string(runtimedebug.Stack())
+		msg += "\nObject created by " + string(runtimedebug.Stack())
 	}
 
-	// set finalizer on obj, not token
-	runtime.SetFinalizer(obj, func(obj any) {
-		// this closure has to use only obj argument and captured "stack" variable
+	// Delete token from map and profile to avoid leaks.
+	c := runtime.AddCleanup(obj, func(msg string) {
+		cleanupByToken.Delete(token)
 
-		msg := fmt.Sprintf("%T has not been finalized", obj)
-		if stack != "" {
-			msg += "\nObject created by " + stack
+		if p := pprof.Lookup(name); p != nil {
+			p.Remove(token)
 		}
```

Comment:
I'm not sure why we do that. We use this package precisely to track leaks. We don't want to avoid them implicitly there.

Reply:
Understood, I haven't to looked with this eye. Addressed.